File: problem200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        count=0
        maxR=len(grid)
        maxC=len(grid[0])
        def countIslands(r, c):
            nonlocal grid, count, maxR, maxC
            if r<0 or r>=maxR or c<0 or c>=maxC or grid[r][c]!="1":
                return
            grid[r][c]="-1"
            countIslands(r-1,c)
            countIslands(r+1,c)
            countIslands(r,c+1)
            countIslands(r,c-1)

        for i in range(maxR):
            for j in range(maxC):
                #print(i,j, grid[i][j])
                if grid[i][j]=="1":
                    count+=1
                    countIslands(i,j)

        return count


#time: O(MN)
#Space: O(MN)

File: problem200.cs
public class Solution {
    char[][] grid;
    int count;
    public int NumIslands(char[][] grid) {
        count=0;
        this.grid = grid;
        
        for(int i=0; i<grid.Length; i++)
        {
            for(int j=0; j<grid[0].Length; j++)
            {
                if(grid[i][j]=='1')
                {
                    count++;
                    Dfs(i,j);
                }
            }
        }
        
        return count;
    }
    
    public void Dfs(int row, int col)
    {
        
        grid[row][col]='2';
        
        if((row+1)<grid.Length && (col)<grid[0].Length && grid[row+1][col]=='1')
            Dfs(row+1, col);
        
        if((row)<grid.Length && (col+1)<grid[0].Length && grid[row][col+1]=='1')
            Dfs(row, col+1);
        
        if((row-1)>=0 && (col)<grid[0].Length && grid[row-1][col]=='1')
            Dfs(row-1, col);
        
        if((row)<grid.Length && (col-1)>=0 && grid[row][col-1]=='1')
            Dfs(row, col-1);
    }
}

//time: O(MN)
//space: O(MN)